vpn/dialer.go

```diff
@@ -96,32 +117,29 @@ func (td TunDialer) DialTimeout(network, address string, timeout time.Duration)
 
 // DialContext connects to the address on the named network using
 // the provided context.
-func (td TunDialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
-	tnet, err := td.createNetTUN(ctx)
-	if err != nil {
-		return nil, err
+func (td *TunDialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	td.mu.Lock()
+	defer td.mu.Unlock()
+	if td.tun == nil {
+		tnet, err := td.createNetTUN(ctx)
+		if err != nil {
+			return nil, err
+		}
+		td.tun = tnet
 	}
-	return tnet.DialContext(ctx, network, address)
+	return td.tun.DialContext(ctx, network, address)
 }
 
-func (td TunDialer) createNetTUN(ctx context.Context) (*netstack.Net, error) {
-	if td.Dialer != nil {
-		td.raw.dialer = td.Dialer
-	}
-
-	client, err := td.raw.dial(ctx)
-	if err != nil {
-		return nil, err
-	}
-	localIP := client.LocalAddr().String()
+func (td *TunDialer) createNetTUN(ctx context.Context) (*netstack.Net, error) {
+	localIP := td.client.LocalAddr().String()
 
 	// create a virtual device in userspace, courtesy of wireguard-go
 	tun, tnet, err := netstack.CreateNetTUN(
 		[]netip.Addr{netip.Addr(netip.MustParseAddr(localIP))},
 		[]netip.Addr{
 			netip.MustParseAddr(td.ns1),
 			netip.MustParseAddr(td.ns2)},
-		client.tunnel.mtu-100,
+		td.client.tunInfo.mtu-100,
 	)
 	// BUG(ainghazal): we cannot use the tun-mtu that the remote
```

Comment:
(very minor) I suggest to use `TODO(name)` for everything ranging from bugs to TODO, with an emphasis on mentioning that a bug is something to be fixed. It simplifies grepping for open problems. Also, if there is already an open issue about this problem, I recommend using `TODO(github_issue_url)`.